### DISCUSSION TITLE
Golang user-defined functions

### DISCUSSION THREAD

**Discussion by joshuamorris3** [cue-lang/cue#2498] [Q&A] [answered]
Besides the [builtin functions](https://pkg.go.dev/cuelang.org/go/pkg) available from the cuelang package, is it possible to call Golang functions from user-defined or other standard library packages. Specifically I would like to use `mail.ParseAddress` from https://pkg.go.dev/net/mail.

**Comment by myitcv** [comment, marked as answer]
This sort of extensibility will be covered by the WebAssembly (Wasm) work that we are doing: https://github.com/cue-lang/cue/discussions/2463. Such Wasm modules will then be distributed via modules (https://github.com/cue-lang/cue/discussions/2330). In this way, people will not be limited by the standard library of builtins that we (currently) offer.

That said, there are often strong arguments for making something part of the standard library. I would suggest that `mail.ParseAddress` comes pretty close to passing an as-yet-undefined threshold for inclusion. Please can you raise a [feature request](https://github.com/cue-lang/cue/issues/new/choose) to that end if you would like this added? It feels ok to me, but a feature request will help to confirm.

(Incidentally, we also somewhat envisage a Go-like `cuelang.org/x/...` namespace of modules for "things that aren't quite standard library but are still useful". That's intentionally a very loose category and definition, but something of a recognition that the import path for a module (with Wasm modules) matters from a provenance perspective). 

cc @4ad for more information on the Wasm side of things.

**Reply by thatInfrastructureGuy** [reply]
@myitcv since [WASM development is paused](https://github.com/cue-lang/cue/issues/2035#issuecomment-2112892919), is there a path forward for user-defined Golang functions?

**Reply by myitcv** [reply]
@thatInfrastructureGuy - we have, in effect, let this point drop through the cracks. Thank you for pointing it out. I have it on my TODO list to ensure that we create a new issue that pulls together the various threads of discussion, issues etc, with a clear statement of where we are (including with respect to WASM), and given the current state of WASM work what our plan b could/should be.

**Reply by myitcv** [reply]
Just bumping with a response here to confirm we are actively working on providing a substantive update here.

**Reply by verdverm** [reply]
The thought crossed my mind that another plan b would be the "workflows in core evaluator" idea, where the evaluator is smart enough to detect and order things appropriately ( I recall this concept being discussed on a community call some time ago...)
 
 I think it is spiritually similar to wasm, just a different "backend" implementation. Users could write "custom functions" that leverage `exec.Exec` to implement using pretty much anything
 
One long desired feature is the "function syntax", where we can write `val: foo(...)` (originally for wrapping #Defs) but could have plain CUE, wasm, and/or workflows; such that the caller doesn't know or care.

**Reply by IngwiePhoenix** [reply]
Instead of opening a new ticket/discussion, I decided to poke in here.
Looking at the docs, I saw references to `ExternalInterpreter` and the WASM module - both of them are obviously still WIP from what I gather.
But after digging my way through tickets and discussions, I haven't really come to a conclusion as to what the current situation here is... would you mind pointing me where I would look for such info? :)
Thank you!

(There are two things I want to try out; using an external scripting language like Goja or seing if I can somehow define "scripts" as a type... but since the latter basically involves adding new AST stuff and alike, I would first want to play around with the former.)

**Comment by joshuamorris3** [comment]
Thank you. I've opened a feature request here https://github.com/cue-lang/cue/issues/2499

**Comment by myitcv** [comment]
Re-opening this discussion given https://github.com/cue-lang/cue/discussions/2498#discussioncomment-9734094 (because I keep trying to find this discussion and failing because it's closed!)

**Comment by gotwarlost** [comment]
Tangentially related in that this should never be a native builtin - I have really, really wanted to match Kubernetes labels against a label selector (using the labels package in k8s). This would really simplify our [cue-crossplane compositions](https://github.com/crossplane-contrib/function-cue) by being able to conditionally set attributes based on object labels. Right now, we have to do some preprocessing elsewhere and hand the results to cue.

**Comment by rajikak** [comment]
Hello, 

Are you working to enable the ability to use, user defined functions ? I see an update in https://github.com/cue-lang/cue/discussions/2498#discussioncomment-9734094, but not sure if you are referring to adding  `mail.ParseAddress` to the standard library or adding the support for enabling user defined functions/libraries. 

I am working on a DSL based on CUE language. A connection to a SFTP server is defined as: 

```
#Connection: {
	Protocol:  "SFTP" | "FTP" | "FTPS" | "SCP" | "SSH"
	Host:      string 
	Port:      int 
	Username?: string
	Password?: string
}
```
so I can declare a connection as:

```
sftpConn: #Connection & {
        Protocol:  "SFTP"
	Host:      "localhost" 
	Port:      22 
	Username: "username"
	Password: "password"
}
```

This works, but I don't want sensitive details go into the version control. So I would like to read `username` and `password` from the environment. So I defined another type as:

```
#EnvParam: { 	
	Name: string
}
```
With that I updated `#Connection` as: 

```
#Connection: {
	Protocol:   "SFTP" | "FTP" | "FTPS" | "SCP" | "SSH"
	Host:      string 
	Port:      int 
	Username?: string | #EnvParam
	Password?: string | #EnvParam
}
```

and I can declare a connection without sensitive details: 

```
sftpConn: #Connection & {
        Protocol:  "SFTP"
	Host:      "localhost" 
	Port:      22 
	Username: #EnvParam & {
            Name:  "SFTP_USERNAME"
        }
	Password: #EnvParam & {
             Name:  "SFTP_PASSWORD"
        }
}
```

This works, but I am thinking if I can import a user defined library this can be more readable:
```
import "userlib"

sftpConn: #Connection & {
        Protocol:  "SFTP"
	Host:      "localhost" 
	Port:      22 
	Username:  userlib.GetEnv("SFTP_USERNAME")
	Password:  userlib.GetEnv("SFTP_PASSWORD")
        }
}
```

Thanks for your work in CUE language.

**Reply by mvdan** [reply]
You can accomplish this with `tool/os.Getenv` and `cue cmd`, as you can see here: https://github.com/cue-lang/cue/blob/f2775f863f4cdeea888aa9cca4ee7273a6f2d2f1/cmd/cue/cmd/testdata/script/cmd_env.txtar

We don't have examples in the rendered docs, but the docs are at https://pkg.go.dev/cuelang.org/go/pkg/tool/os too.

**Reply by rajikak** [reply]
Thanks for the quick reply. I am aware of `tool` package. I am interested in knowing in how can I import my own libraries, so that I can reuse work from other Go libraries/packages (I am using CUE Go API). For example https://pkg.go.dev/github.com/zclconf/go-cty/cty/function/stdlib (https://github.com/hashicorp/terraform/blob/main/internal/lang/functions.go#L64-L180).

**Reply by mvdan** [reply]
Calling your own Go functions from CUE is currently not possible. We continue to have plans in this space but we are currently focused on the new evaluator and performance (https://github.com/cue-lang/cue/issues/2850). If `cue cmd` is not a possibility for you, then the best option right now is to write a Go program combining your own code with the use of our Go APIs at https://pkg.go.dev/cuelang.org/go/cue.

**Reply by aep** [reply]
but the golang api doesnt actually allow defining custom functions either? the api is for that is marked internal

**Reply by myitcv** [reply]
@aep - @mvdan's response was referring to the tooling layer. It sounds like you have requirements with custom functions in the evaluator?

**Comment by pgeiem** [comment]
I am new to cue. I would like to use it as a configuration source for my go application. I need to manipulate lots of dates and dates recurrences I have some go code to handle this and was hoping to build a dedicated package to expose these set of date manipulation functions to cue. I understand here that this is not possible, do I understand right ? Is there really no possibly to add (custom) function as the default built-in ?

**Reply by verdverm** [reply]
Currently no, there would be issues that your code would not be usable by others.

There are efforts around WASM and something more Go specific that can avoid WASM aiui.

Another option in the near term is proposing to expand the available functions in https://pkg.go.dev/cuelang.org/go@v0.10.0/pkg/time  These are generally a proxy to the underlying packages in Go's stdlib
